paddle/fluid/operators/matmul_op.h

```diff
@@ -72,23 +73,57 @@ Tensor CombineBatchAndM(const Tensor& input) {
 // (Warning: This requires transposing data and writes into new memory.)
 // Identity op if the tensor is not of rank 3.
 template <typename DeviceContext, typename T>
-Tensor CombineBatchAndN(const DeviceContext& context, const Tensor& input) {
-  Tensor output;
+inline framework::Tensor CombineBatchAndN(const DeviceContext& context,
+                                          const framework::Tensor& input) {
   auto in_dims = input.dims();
-  if (in_dims.size() == 3) {
-    output.Resize({in_dims[1], in_dims[0], in_dims[2]});
-    output.mutable_data<T>(context.GetPlace());
-    std::vector<int> axis = {1, 0, 2};
-    math::Transpose<DeviceContext, T, 3> trans;
-    trans(context, input, &output, axis);
-    std::vector<int64_t> out_dims = {in_dims[1], in_dims[0] * in_dims[2]};
-    output.Resize({in_dims[1], in_dims[0] * in_dims[2]});
-  } else {
-    output.ShareDataWith(input);
+  if (in_dims.size() != 3) {
+    return input;
   }
+  framework::Tensor output;
+  output.Resize({in_dims[1], in_dims[0], in_dims[2]});
+  output.mutable_data<T>(context.GetPlace());
+  std::vector<int> axis = {1, 0, 2};
+  math::Transpose<DeviceContext, T, 3> trans;
+  trans(context, input, &output, axis);
+  output.Resize({in_dims[1], in_dims[0] * in_dims[2]});
+
   return output;
 }
 
+inline void NormalizeTensorShape(framework::Tensor* x,
+                                 const math::MatDim& mat_dim_x) {
+  int64_t h, w;
+  h = mat_dim_x.height_;
+  w = mat_dim_x.width_;
+  if (mat_dim_x.trans_) {
+    std::swap(w, h);
+  }
+  if (mat_dim_x.batch_size_) {
+    x->Resize({mat_dim_x.batch_size_, h, w});
+  } else {
+    x->Resize({h, w});
+  }
+}
+
+inline void NormalizeXYOutTensorShape(framework::Tensor* x,
```

Comment:
I can hardly invent a name to describe the very complicated operation that's done by this function ...

I would highly recommend hiding the definition of this function in a nested namespace or in the class template MatMulGradKernel -- just not to make it pollute the namespace paddle.fluid.operator, so could we revisit its name later when we have a better idea.

Reply:
Renamed to `ReshapeXYOutIntoMatrixSequence`.

BTW, I found the header `matmul_op.h` is not necessary, I just removed it and moving the implementations into `matmul_op.cc`. The definitions are hiding in the source file and no symbol is exported.